File: striver sheet/recursion/46. Permutations.py

import copy
class Solution:
    

    def permute(self, nums: List[int]) -> List[List[int]]:
        def solve(freq,ds,ans,nums):
            if(len(ds)==len(nums)):
                ans.append(deepcopy(ds))
                return
            for i in range(len(nums)):
                if not freq[i]:
                    freq[i]=1
                    ds.append(nums[i])
                    solve(freq,ds,ans,nums)
                    freq[i]=0
                    ds.pop()
        ds=[]
        ans=[]
        freq=[0]*len(nums)
        solve(freq,ds,ans,nums)
        return ans

File: striver sheet/recursion/46. Permutations.cpp
class Solution {
public:
void solve(vector<int>&nums,vector<int>&ds,vector<int>&freq,vector<vector<int>>&ans){
    if(ds.size()==nums.size()){
        ans.push_back(ds);
        return ;
    }
    for(int i=0;i<nums.size();i++){
        if(!freq[i]){
            freq[i]=1;
            ds.push_back(nums[i]);
            solve(nums,ds,freq,ans);
            freq[i]=0;
           ds.pop_back();
       
        }
    }
}
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>>ans;
        vector<int>ds;
        vector<int>freq(nums.size(),0);
        
        solve(nums,ds,freq,ans);
        return ans;
    }
};